### PR TITLE
[BEAM-1401] Allow unbounded windowed PCollections for FileBasedSinks

### PR DESCRIPTION
**Pull request by reuvenlax**

### CODE REVIEW

**Comment by reuvenlax** [comment]
retest this please

**Comment by kennknowles** [comment]
R: @kennknowles

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 1]
```diff
@@ -293,6 +402,9 @@ private static String getFileExtension(String usersExtension) {
     /** Directory for temporary output files. */
     protected final ValueProvider<String> tempDirectory;
 
+    /** Whether to delete the entire temp directory in finalize. */
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: javadoc doesn't quite match the field

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 2]
```diff
@@ -135,21 +148,113 @@ public String getMimeType() {
...
+      implements org.apache.beam.sdk.transforms.SerializableFunction<
+       FilenamePolicy.Context, String> {
+    /**
+     * Context used for generating a name based on window, pane, shard numer, and num shards.
+     * Window and pane will only be provided if windowed writes have been requested using
+     * {@link Write.Bound#withWindowedWrites}. The policy must produce unique filenames for unique
+     * Context objects.
+     */
+    public static class Context {
+      private BoundedWindow window;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: Mark `@Nullable` those things that might be null.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java [hunk 3]
```diff
@@ -649,22 +672,39 @@ public Bound withFooter(@Nullable String footer) {
       public Bound withWritableByteChannelFactory(
           WritableByteChannelFactory writableByteChannelFactory) {
         return new Bound(name, filenamePrefix, filenameSuffix, header, footer, numShards,
-            shardTemplate, validate, writableByteChannelFactory);
+            shardTemplate, validate, writableByteChannelFactory, filenamePolicy, windowedWrites);
+      }
+
+      public Bound withWindowedWrites() {
+        return new Bound(name, filenamePrefix, filenameSuffix, header, footer, numShards,
+            shardTemplate, validate, writableByteChannelFactory, filenamePolicy, true);
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java]: I may have missed it, but shouldn't the default template change to be window+pane aware?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTest.java [hunk 4]
```diff
@@ -275,6 +290,119 @@ public void testAvroIOWriteAndReadSchemaUpgrade() throws Throwable {
...
+
+    @Override
+    public ValueProvider<String> getBaseOutputFilenameProvider() {
+      return StaticValueProvider.of(outputFilePrefix);
+    }
+
+    @Override
+    public String apply(Context input) {
+      String filename = outputFilePrefix + "-" + input.getWindow().toString() +  "-"
+          + input.getShardNumber() + "-of-" + (input.getNumShards() - 1) + "-pane-"
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTest.java]: Commented elsewhere, but it seems like we should provide this or something like it as a default, no?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java [hunk 5]
```diff
@@ -241,6 +248,12 @@ public abstract void finalize(Iterable<WriteT> writerResults, PipelineOptions op
      */
     public abstract void open(String uId) throws Exception;
 
+    public abstract void setWindowAndPane(BoundedWindow window, PaneInfo paneInfo) throws Exception;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: I'm not arguing against it or anything, but these setter methods seem incongruous relative to the rest of the API here. Are writers being reused in a novel way? (From what I can tell this just sets up the APIs that runners will later implement). Two other natural places to pass this info would be via the `FileBasedWriteOperation` and the call to `open`. I'm trying to grok the choice.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java [hunk 6]
```diff
@@ -540,6 +565,8 @@ private Read() {}
         this.schema = schema;
         this.validate = validate;
         this.codec = codec;
+        this.windowedWrites = windowedWrites;
+        this.filenamePolicy = filenamePolicy;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java]: Can you convert prefix/suffix into just a `FilenamePolicy`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 7]
```diff
@@ -384,22 +502,55 @@ public void initialize(PipelineOptions options) throws Exception {}
...
+      List<String> files = new ArrayList<>();
+      for (FileResult result : writerResults) {
+        if (result.getDestinationFilename() != null) {
+          outputFilenames.put(result.getFilename(), result.getDestinationFilename());
+        } else {
+          files.add(result.getFilename());
+        }
+      }
+
+      // If the user does not specify numShards() (not supported with windowing). Then the
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: It took me a minute to figure out this comment. So this whole code path is dead in streaming, hence it is OK to hard-code the fact that no window or pane info is passed?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java (hunk 2 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java (hunk 1 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: Fixed

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java (hunk 7 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: Yes

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java (hunk 3 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java]: Right now no, you must specify a Strategy. If we want default templates for this, I think that's a valid feature request that should go into a different PR

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTest.java (hunk 4 above)
Comment (reuvenlax) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTest.java]: I didn't, because I'm not sure it's the default everyone wants :) If we do want default pattern-based support for this, I'd be in favor of doing it as a followon PR (it's backwards compatible, because it can always be a new Policy object

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java (hunk 5 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: Moved into open()

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java (hunk 6 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java]: Right now they get translated into DefaultFilenamePolicy. We could refactor further and force the user to specify the policy themselves, however I think that's best left to a subsequent PR

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java [hunk 8]
```diff
@@ -255,12 +255,12 @@ public void processElement(ProcessContext c, BoundedWindow window) throws Except
...
-          if (windowedWrites) {
-            writer.setWindowAndPane(window, c.pane());
-          } else {
-            writer.setWindowAndPane(null, null);
-          }
+
+          writer.open(UUID.randomUUID().toString(),
+              windowedWrites ? window : null,
+              windowedWrites ? c.pane() : null,
+              -1,
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java]:
```java
public static final int UNKNOWN_NUMSHARDS = -1;
public static final int UNKNOWN_SHARDNUM = -1
```

or bite the bullet and box them

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java [hunk 9]
```diff
@@ -237,7 +239,16 @@ public HDFSWriter(HDFSWriteOperation<K, V> writeOperation,
     }
 
     @Override
-    public void open(String uId) throws Exception {
+    public final void open(String uId,
+                           @Nullable BoundedWindow window,
+                           @Nullable PaneInfo paneInfo,
+                           int shard,
+                           int numShards) throws Exception {
+      if (window != null) {
```
Comment (kennknowles) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java]: Just to be extra safe, in case there's a bug in the caller, might as well `window != null || paneInfo != null`

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java [hunk 10]
```diff
@@ -237,7 +239,16 @@ public HDFSWriter(HDFSWriteOperation<K, V> writeOperation,
...
+        throw new UnsupportedOperationException("Windowing support not implemented yet for"
```
Comment (kennknowles) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java]: I think it would be a bit more actionable to indicate what window was passed, here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java [hunk 11]
```diff
@@ -500,21 +502,28 @@ public void processElement(ProcessContext c) throws Exception {
...
+                  .add(writeOperationView)
+                  .add(numShardsView);
+          results  = input
               .apply("ApplyShardLabel", ParDo.of(
-                  new ApplyShardingKey<T>(numShardsView, null)).withSideInputs(numShardsView));
+                  new ApplyShardingKey<T>(numShardsView, null)).withSideInputs(numShardsView))
+              .apply("GroupIntoShards", GroupByKey.<Integer, T>create())
+              .apply("WriteShardedBundles",
+                  ParDo.of(new WriteShardedBundles<>(writeOperationView, numShardsView))
+                      .withSideInputs(sideInputs.build()));
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java]: `.withSideInputs` accepts varags, if you'd like to save a few lines. The variable isn't reused, is it?

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/simpleauth/SimpleAuthHDFSFileSink.java [hunk 12]
```diff
@@ -99,18 +102,23 @@ public SimpleAuthHDFSWriter(SimpleAuthHDFSWriteOperation<K, V> writeOperation,
     }
 
     @Override
-    public void open(final String uId) throws Exception {
+    public void open(String uId,
+                     @Nullable BoundedWindow window,
+                     @Nullable PaneInfo paneInfo,
+                     int shard,
+                     int numShards) throws Exception {
+      final String id = uId;
```
Comment (kennknowles) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/simpleauth/SimpleAuthHDFSFileSink.java]: checkstate / unsupport operation exception if window or pane non-null?

**Comment by reuvenlax** [comment]
retest this please

**Comment by kennknowles** [comment]
retest this please

**Comment by reuvenlax** [comment]
retest this please

**Comment by kennknowles** [comment]
Getting this patched up. There's pretty epic conflicts (all easy) due to style guide driven revision.

**Comment by reuvenlax** [comment]
What's the status of this?

**Comment by kennknowles** [comment]
I took these commits, rebased them and resolved the conflicts and squashed them. I then did a little fixup to use it in the example in a different way. This is all in #2212. I didn't push the tests over the finish line, though the failures look unrelated at first glance. I will be at Strata SJC this week so probably won't make much progress. If you want to push it further I recommend starting from that point (if you follow our contributor guide, this is `git reset --hard github/pr/2212`) so you don't have to resolve conflicts.

**Comment by reuvenlax** [comment]
@kennknowles Reset to your PR, and changed the interface to more cleanly separate between the windowed and unwindowed cases.

**Comment by reuvenlax** [comment]
retest this please

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 13]
```diff
@@ -146,21 +160,150 @@ public String getMimeType() {
...
+        this.extension = extension + getFileExtension(
+            writableByteChannelFactory.getFilenameSuffix());
+      } else {
+        this.extension = extension;
+      }
+      this.fileNamingTemplate = fileNamingTemplate;
+    }
+
+    @Override
+    public String unwindowedFilename(FilenamePolicy.Context context) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: Just inline the few params from the contexts. Otherwise you still have the problem of adding information to the context that are required to be handled and some old policy doesn't notice. We pretty explicitly don't want smooth compatibility here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java [hunk 14]
```diff
@@ -238,8 +246,18 @@ public abstract void finalize(Iterable<WriteT> writerResults, PipelineOptions op
      * <p>The unique id that is given to open should be used to ensure that the writer's output does
      * not interfere with the output of other Writers, as a bundle may be executed many times for
      * fault tolerance. See {@link Sink} for more information about bundle ids.
+     *
+     * <p></p>The window and paneInfo arguments are populated when windowed writes are requested.
+     * shard and numbShards are populated for the case of static sharding. In cases where the
+     * runner is dynamically picking sharding, shard and numShards might both be set to -1.
      */
-    public abstract void open(String uId) throws Exception;
+    public abstract void open(String uId,
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java]: I was expecting to see `writer.openGlobal(base, shardnum, numshards)` and `writer.openWindowed(base, window, pane, shardnum, numshards)`. This way, a writer either supports windowing or not based on the method, not nulls.

Incidentally, why does `open` not just take an already-formatted filename?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/FileIOChannelFactory.java [hunk 15]
```diff
@@ -175,17 +177,20 @@ public Path toPath(String path) {
...
 
   @Override
-  public void copy(List<String> srcFilenames, List<String> destFilenames) throws IOException {
+  public void copy(Iterable<String> srcFilenames, Iterable<String> destFilenames) throws
+      IOException {
+    List<String> srcList = Lists.newArrayList(srcFilenames);
+    List<String> destList = Lists.newArrayList(destFilenames);
     checkArgument(
-        srcFilenames.size() == destFilenames.size(),
+        srcList.size() == destList.size(),
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/FileIOChannelFactory.java]: These refactors could be broken out and submitted any time.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java [hunk 16]
```diff
@@ -635,23 +636,27 @@ public Void call() throws IOException {
     return batches;
   }
 
-  public void copy(List<String> srcFilenames, List<String> destFilenames) throws IOException {
+  public void copy(Iterable<String> srcFilenames,
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java]: ditto on submitted prerequisite refactors

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java [hunk 17]
```diff
@@ -302,23 +365,32 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
+    private final PCollectionView<Integer> numShardsView;
+    private final ValueProvider<Integer> numShardsProvider;
     private int shardNumber;
 
-    ApplyShardingKey(PCollectionView<Integer> numShards) {
-      this.numShards = numShards;
-      shardNumber = -1;
+    ApplyShardingKey(PCollectionView<Integer> numShardsView,
+                     ValueProvider<Integer> numShardsProvider) {
+      this.numShardsView = numShardsView;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java]: Refactors to how sharding is passed around could submit early, too

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java [hunk 18]
```diff
@@ -396,133 +470,133 @@ public void processElement(ProcessContext c) throws Exception {
...
     // There is a dependency between this ParDo and the first (the WriteOperation PCollection
     // as a side input), so this will happen after the initial ParDo.
     PCollection<WriteT> results;
-    final PCollectionView<Integer> numShards;
-    if (computeNumShards == null) {
-      numShards = null;
-      results =
-          inputInGlobalWindow.apply(
-              "WriteBundles",
+    final PCollectionView<Integer> numShardsView;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java]: Somewhere in `expand` (if not compile time) we should check that the filename policy supports windowing OR the input is globally windowed.

And I guess part of the challenge is that `Write` doesn't know it has a `FileBasedSink`, so `Sink` will have to surface something. We can follow up by inlining them...

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java (hunk 13 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: As discussed, added a comment.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java (hunk 18 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java]: As discussed, will happen when we get rid of Sink

**Comment by kennknowles** [comment]
I'm thinking of `-XX:MaxJavaStackTraceDepth` needs to be cranked up or `-1` even.

**Comment by kennknowles** [comment]
retest this please

File: examples/java/pom.xml [hunk 19]
```diff
@@ -209,7 +209,6 @@
                 <configuration>
                   <includes>
                     <include>WordCountIT.java</include>
-                    <include>WindowedWordCountIT.java</include>
```
Comment (kennknowles) [inline examples/java/pom.xml]: Filed https://issues.apache.org/jira/browse/BEAM-1886

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 20]
```diff
@@ -610,8 +796,37 @@ protected void finishWrite() throws Exception {}
...
+    public final void openUnwindowed(String uId,
+                                     int shard,
+                                     int numShards) throws Exception {
+      if (getWriteOperation().windowedWrites) {
+        throw new IllegalStateException("openUnwindowed called a windowed sink.");
+      }
+      open(uId, null, null, shard, numShards);
+    }
+
+    private void open(String uId,
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: Noting that after inlining things, the full filename should be owned and computed outside of here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java [hunk 21]
```diff
@@ -653,8 +875,17 @@ public final FileResult close() throws Exception {
           throw new IllegalStateException("Channel should only be closed by its owner: " + channel);
         }
       }
-      FileResult result = new FileResult(filename);
-      LOG.debug("Result for bundle {}: {}", this.id, filename);
+
+      FilenamePolicy filenamePolicy = getWriteOperation().getSink().fileNamePolicy;
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: The writer should just receive the filename that was computed outside.

**Comment by kennknowles** [comment]
OK LGTM with followups on the way.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java (hunk 20 above)
Comment (reuvenlax) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java]: Don't think so, because in the case of dynamic sharding (which will be more full supported in the followup PR), the filename cannot be calculated until finalize.
